### PR TITLE
Add lint for duplicate methods of trait bounds

### PR DESCRIPTION
**Pull request by wiomoc**
rel: #5777

changelog: Add [`trait_duplication_in_bounds`] lint

### CODE REVIEW

File: clippy_lints/src/trait_bounds.rs [hunk 1]
```diff
@@ -29,6 +30,35 @@ declare_clippy_lint! {
...
+    /// fn func<T: Clone + Default>(arg: T) {}
+    /// ```
+    /// or
+    ///  ///
+    /// ```rust
+    /// fn func<T>(arg: T) where T: Clone + Default {}
+    /// ```
+    pub TRAIT_DUPLICATION_IN_BOUNDS,
+    pedantic,
+    "Check if the same trait bounds are specifed twice during a function declaration"
```
Comment (giraffate) [inline clippy_lints/src/trait_bounds.rs:59]:
```suggestion
    "Check if the same trait bounds are specified twice during a function declaration"
```

File: clippy_lints/src/trait_bounds.rs [hunk 2]
```diff
@@ -101,3 +146,48 @@ impl<'tcx> LateLintPass<'tcx> for TraitBounds {
         }
     }
 }
+
+fn check_trait_bound_duplication(cx: &LateContext<'_>, gen: &'_ Generics<'_>) {
+    if in_macro(gen.span) {
+        return;
+    }
+
+    let mut map = FxHashMap::default();
```
Comment (Manishearth) [inline clippy_lints/src/trait_bounds.rs:155]: we should early return if:

 - there are no parameters
 - there are no predicates in the where clause

**Comment by flip1995** [comment]
A nice thing to do here would be to add a suggestion, that can auto-remove those bounds. But this involves some effort, so it should can be done in another PR.

**Comment by Manishearth** [comment]
@bors r+

@flip1995 yeah i was going to suggest this, unfortunately suggestions removing a bound require some weird spans where you include one of the `+` operators, and it's icky.

**Comment by flip1995** [comment]
Yeah, I would probably do this by modifying the ast and pretty printing it

**Comment by flip1995** [comment]
Blocked on #5864

**Pull request merged**